pkg/container-hook/tracer_test.go

```diff
@@ -0,0 +1,130 @@
+					Type:        containerhook.EventTypeAddContainer,
+					ContainerID: containerID,
+				}
+			}),
+		},
+		"one_container_after_some_failed_containers": {
+			generateEvent: generateEvent(2),
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, containerID string) *containerhook.ContainerEvent {
+				return &containerhook.ContainerEvent{
+					Type:        containerhook.EventTypeAddContainer,
+					ContainerID: containerID,
+				}
+			}),
+		},
+		"one_container_after_many_failed_containers": {
+			// Test with a number bigger than FANOTIFY_DEFAULT_MAX_GROUPS
+			// https://github.com/torvalds/linux/blob/v6.9/fs/notify/fanotify/fanotify_user.c#L32
+			// #define FANOTIFY_DEFAULT_MAX_GROUPS	128
+			generateEvent: generateEvent(3), // TODO: change this number to 130
+			validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, containerID string) *containerhook.ContainerEvent {
+				return &containerhook.ContainerEvent{
+					Type:        containerhook.EventTypeAddContainer,
+					ContainerID: containerID,
+				}
+			}),
+		},
+	} {
+		test := test
+
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			events := []containerhook.ContainerEvent{}
+			eventCallback := func(event containerhook.ContainerEvent) {
+				// normalize
+				event.ContainerName = ""
+				event.ContainerPID = 0
+				event.ContainerConfig = nil
+				event.Bundle = ""
+
+				events = append(events, event)
+			}
+
+			notifier, err := containerhook.NewContainerNotifier(eventCallback)
+			require.Nil(t, err, "Returned error was not nil")
+			require.NotNil(t, notifier, "Returned notifier was nil")
+
+			containerID, err := test.generateEvent()
+			require.Nil(t, err, "Returned error was not nil")
+
+			// Give some time for the tracer to capture the events
+			time.Sleep(100 * time.Millisecond)
+
+			test.validateEvent(t, nil, containerID, events)
+		})
+	}
+}
+
+func generateEvent(nfailure int) func() (string, error) {
+	return func() (string, error) {
+		// Failed container with bad configuration
+		for i := 0; i < nfailure; i++ {
+			cmd := exec.Command("docker", "run", "--rm", "-v", "/dev/null:/dev/null/invalid", "busybox", "id")
+			stdoutStderr, err := cmd.CombinedOutput()
+			trimedOutput := strings.TrimSuffix(string(stdoutStderr), "\n")
+			if err == nil {
+				return trimedOutput, errors.New("container with invalid configuration should have failed")
+			}
+		}
+
+		// Good container with valid configuration
+		var stdout, stderr bytes.Buffer
+		cmd := exec.Command("docker", "run", "--rm", "-d", "busybox", "id")
+		cmd.Stdout = &stdout
+		cmd.Stderr = &stderr
+		err := cmd.Run() // Avoid CombinedOutput because stderr might contain busybox pulling messages
+		if err != nil {
+			return "", fmt.Errorf("running container: %w\nStdout:\n%s\nStderr:\n%s",
+				err, string(stdout.Bytes()), string(stderr.Bytes()))
+		}
+		trimedOutput := strings.TrimSuffix(string(stdout.Bytes()), "\n")
+		return trimedOutput, nil
+	}
```

Comment:
Perhaps we can use a call to the API instead of invoking the CLI?

We already have a library for it, I gave it a try:

```diff
diff --git a/pkg/container-hook/tracer_test.go b/pkg/container-hook/tracer_test.go
index b15b025d4..317c82067 100644
--- a/pkg/container-hook/tracer_test.go
+++ b/pkg/container-hook/tracer_test.go
@@ -18,11 +18,8 @@
 package containerhook_test

 import (
-       "bytes"
-       "errors"
        "fmt"
-       "os/exec"
-       "strings"
+       "math/rand"
        "testing"
        "time"

@@ -30,13 +27,14 @@ import (

        utilstest "github.com/inspektor-gadget/inspektor-gadget/internal/test"
        containerhook "github.com/inspektor-gadget/inspektor-gadget/pkg/container-hook"
+       "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/testutils"
 )

 func TestContainerHook(t *testing.T) {
        utilstest.RequireRoot(t)

        type testDefinition struct {
-               generateEvent func() (string, error)
+               generateEvent func(t *testing.T) string
                validateEvent func(t *testing.T, info *utilstest.RunnerInfo, stdoutStderr string, events []containerhook.ContainerEvent)
        }

@@ -63,7 +61,7 @@ func TestContainerHook(t *testing.T) {
                        // Test with a number bigger than FANOTIFY_DEFAULT_MAX_GROUPS
                        // https://github.com/torvalds/linux/blob/v6.9/fs/notify/fanotify/fanotify_user.c#L32
                        // #define FANOTIFY_DEFAULT_MAX_GROUPS  128
-                       generateEvent: generateEvent(3), // TODO: change this number to 130
+                       generateEvent: generateEvent(130), // TODO: change this number to 130
                        validateEvent: utilstest.ExpectAtLeastOneEvent(func(info *utilstest.RunnerInfo, stdoutStderr string) *containerhook.ContainerEvent {
                                return &containerhook.ContainerEvent{
                                        Type:        containerhook.EventTypeAddContainer,
@@ -91,40 +89,31 @@ func TestContainerHook(t *testing.T) {
                        require.Nil(t, err, "Returned error was not nil")
                        require.NotNil(t, notifier, "Returned notifier was nil")

-                       stdoutStderr, err := test.generateEvent()
-                       require.Nil(t, err, "Returned error was not nil")
+                       containerID := test.generateEvent(t)

                        // Give some time for the tracer to capture the events
                        time.Sleep(100 * time.Millisecond)

-                       test.validateEvent(t, nil, stdoutStderr, events)
+                       test.validateEvent(t, nil, containerID, events)
                })
        }
 }

-func generateEvent(nfailure int) func() (string, error) {
-       return func() (string, error) {
+func generateEvent(nfailure int) func(t *testing.T) string {
+       return func(t *testing.T) string {
                // Failed container with bad configuration
                for i := 0; i < nfailure; i++ {
-                       cmd := exec.Command("docker", "run", "--rm", "-v", "/dev/null:/dev/null/invalid", "busybox", "id")
-                       stdoutStderr, err := cmd.CombinedOutput()
-                       trimedOutput := strings.TrimSuffix(string(stdoutStderr), "\n")
-                       if err == nil {
-                               return trimedOutput, errors.New("container with invalid configuration should have failed")
-                       }
+                       fmt.Printf("running %d\n", i)
+                       name := fmt.Sprintf("ig-%d", rand.Uint32())
+                       // foo is not a valid command inside busybox, hence it fails
+                       container := testutils.NewDockerContainer(name, "foo", testutils.WithoutLogs())
+                       container.Run(t)
                }

                // Good container with valid configuration
-               var stdout, stderr bytes.Buffer
-               cmd := exec.Command("docker", "run", "--rm", "-d", "busybox", "id")
-               cmd.Stdout = &stdout
-               cmd.Stderr = &stderr
-               err := cmd.Run() // Avoid CombinedOutput because stderr might contain busybox pulling messages
-               if err != nil {
-                       return "", fmt.Errorf("running container: %w\nStdout:\n%s\nStderr:\n%s",
-                               err, string(stdout.Bytes()), string(stderr.Bytes()))
-               }
-               trimedOutput := strings.TrimSuffix(string(stdout.Bytes()), "\n")
-               return trimedOutput, nil
+               name := fmt.Sprintf("ig-%d", rand.Uint32())
+               container := testutils.NewDockerContainer(name, "id", testutils.WithoutLogs())
+               container.Run(t)
+               return container.ID()
        }
 }
```

However it doesn't work fine because the test with 130 failed containers passes.

I also tried with RunDockerFailedContainer, but it didn't work.